### ISSUE TITLE
allow negated condition in for-in guards in guard-for-in rule

### ISSUE THREAD

**Issue by michaelficarra** [eslint/eslint#7567]
I would like `guard-for-in` to allow the following pattern in addition to what it currently allows.

```js
for (key in foo) {
    if (!{}.hasOwnProperty.call(foo, key)) continue;
    doSomething(key);
}
```

To be clear, the difference is that this guard has a negated condition with a `continue` as the consequent. The `continue` should be allowed as the only statement in a block as well, to be compatible with that brace style preference.

```js
for (key in foo) {
    if (!{}.hasOwnProperty.call(foo, key)) {
        continue;
    }
    doSomething(key);
}
```

**edit:**

**What rule do you want to change?**

`guard-for-in`

**Does this change cause the rule to produce more or fewer warnings?**

Fewer.

**How will the change be implemented? (New option, new default behavior, etc.)?**

New default behaviour.

**Please provide some example code that this change will affect:**

```js
for (key in foo) {
    if (!{}.hasOwnProperty.call(foo, key)) continue;
    doSomething(key);
}
```

```js
for (key in foo) {
    if (!{}.hasOwnProperty.call(foo, key)) {
        continue;
    }
    doSomething(key);
}
```

**What does the rule currently do for this code?**

Warns.

**What will the rule do after it's changed?**

Not warn.

**Comment by platinumazure** [comment]
Maybe this should be refactored to use code path analysis, so that we could flag the for statement if any path is reachable without having gone through a hasOwnProperty check. Of course, it doesn't help that right now we're not even checking what the IfStatement test looks like...

:+1: from me.

**Comment by platinumazure** [comment]
Oh, @michaelficarra, can you please edit your post to match [this template](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md)? Thanks!

**Comment by michaelficarra** [comment]
@platinumazure Done.

**Comment by not-an-aardvark** [comment]
@michaelficarra Are you still interested in pushing this proposal forward?

**Comment by michaelficarra** [comment]
@not-an-aardvark Yes. What does that involve?

**Comment by not-an-aardvark** [comment]
It generally involves drumming up support from the team (see [here](http://eslint.org/docs/maintainer-guide/issues.html#championing-issues)). Right now there are two :+1:s from team members on the proposal, so we would need one more for the proposal to be accepted.

**Comment by michaelficarra** [comment]
@not-an-aardvark Does my own :+1: count?

**Comment by not-an-aardvark** [comment]
No, it does not (it requires three :+1:s in addition to a champion).

ccing @eslint/eslint-team (sorry about all the pings today) to see if anyone else is in favor of this.

**Comment by platinumazure** [comment]
We don't have a champion on this, so I'm setting this back to evaluating. If we don't get a champion in a week or two, we should probably close this.

**Comment by ilyavolodin** [comment]
@platinumazure Since @michaelficarra created this issue, and filed a PR, I assume he is championing it, so setting it back to accepted.

**Comment by platinumazure** [comment]
@ilyavolodin @michaelficarra Sorry for the mixup!
